code/controllers/subsystem/achievements.dm

```diff
@@ -101,3 +101,33 @@ SUBSYSTEM_DEF(achievements)
 
 	if(to_update.len)
 		SSdbcore.MassInsert(format_table_name("achievement_metadata"),to_update,duplicate_key = TRUE)
+
+	var/list/orphaned_keys = get_orphaned_keys()
+	if(orphaned_keys.len)
+		message_admins("Achievement metadata found without matching achievement, use Achievement-Admin-Panel verb to cleanup if necessary")
+
+// returns metadata keys in db with no matching achievement datum, either deleted achievements, or from server with code ahead of us
+/datum/controller/subsystem/achievements/proc/get_orphaned_keys()
+	. = list()
+	var/list/current_metadata = list()
+	//select metadata here
+	var/datum/db_query/Q = SSdbcore.NewQuery("SELECT achievement_key,achievement_version FROM [format_table_name("achievement_metadata")]")
+	if(!Q.Execute(async = TRUE))
+		qdel(Q)
+		return
+	else
+		while(Q.NextRow())
+			current_metadata[Q.item[1]] = text2num(Q.item[2])
```

Comment:
can you have a comment explaining how this works?

Reply:
This part is actually unnecessary right now, it only needs the keys right now, i'll simplify it in a moment.